src/Components/Server/src/ComponentsHub.cs

```diff
@@ -42,32 +47,97 @@ public ComponentsHub(IServiceProvider services, ILogger<ComponentsHub> logger)
         internal CircuitHost CircuitHost
         {
             get => (CircuitHost)Context.Items[CircuitKey];
-            private set => Context.Items[CircuitKey] = value;
+            set => Context.Items[CircuitKey] = value;
         }
 
+        // For unit testing
+        internal Task OnBeforeDeactivate { get; set; } = Task.CompletedTask;
+        internal Task OnBeforeActivate { get; set; } = Task.CompletedTask;
+
         /// <summary>
         /// Intended for framework use only. Applications should not call this method directly.
         /// </summary>
         public override async Task OnDisconnectedAsync(Exception exception)
         {
-            await CircuitHost.DisposeAsync();
+            var circuitHost = CircuitHost;
+            if (circuitHost == null)
+            {
+                return;
+            }
+
+            CircuitHost = null;
+            try
+            {
+                if (!await AcquireCircuitLock())
+                {
+                    // If we're unable to enter the critical section, clear this instance and throw.
```

Comment:
What are the cases where this could happen?

Reply:
I ended up removing the time limits. It just felt correct to have meaningful limits, but this is pretty identical to having a `lock` statement and we would expect it to enter the critical section eventually.